core/finality-grandpa/src/lib.rs

```diff
@@ -252,6 +252,30 @@ struct TopicTracker {
 	set_id: u64,
 }
 
+impl TopicTracker {
+	fn is_expired(&self, round: u64, set_id: u64) -> bool {
+		if set_id < self.set_id {
+			trace!(target: "afg", "Expired: Message with expired set_id {} (ours {})", set_id, self.set_id);
+			return true;
+		} else if set_id == self.set_id + 1 {
+			// allow a few first rounds of future set.
+			if round > MESSAGE_ROUND_TOLERANCE {
+				trace!(target: "afg", "Expired: Message too far in the future set, round {} (ours set_id {})", round, self.set_id);
```

Comment:
Shouldn't we tag messages within the tolerance as `Future`?

Reply:
I don't think so, because we have enough information to validate them